Tensile/Source/client/source/Reference.cpp

```diff
@@ -292,7 +320,9 @@ namespace Tensile
                                 bVal = Transform<typename Inputs::BType>::Input(
                                     inputs.b[bIndex + (bI * bStride) - zpB.padStart], bConjugate);
 
-                            value += static_cast<Accumulator>(aVal * bVal);
+                            value += multiply<typename Inputs::AType,
```

Comment:
I would suggest declaring a constexpr flag like before. Then add type aliases for multiplicands: this case for A and B.

``` 
/* Transform the data type of A/B to Accumulator if A!=ACC or B!=ACC, but filter out cases, I8/I32/I32 and I8x4/I32/I32
 * 
 * There are three cases of doing multiplication and their conditions to do transform or not are as below.
 * 1. AxB : (A!=ACC or B!=ACC) and A!=I8 and A!=I8x4
 * 2. Alpha x rC :  (Alpha!=ACC or rC!=ACC)
 * 3. Beta x C : (Beta!=ACC or C!=ACC)
 */
  constexpr bool needAccumCast
      = !(std::is_same<typename Inputs::AType, Accumulator>() && std::is_same<Inputs::BType, Accumulator>())
        && !std::is_same<typename Inputs::AType, Int8>() //case I8/I32/I32
        && !std::is_same<typename Inputs::AType, Int8x4>(); //case I8x4/I32/I32 
        
  using AMultT = std::conditional_t<needAccumCast, Accumulator, typename Inputs::AType>;
  using BMultT = std::conditional_t<needAccumCast, Accumulator, typename Inputs::BType>;
   
  value += static_cast<Accumulator>(static_cast<AMultT>(aVal) * static_cast<BMultT>(bVal));
        
```

Reply:
@cgmillette Thanks for your suggestion.
The idea of using the type aliases is helpful, which can be used to instead the enable_if in the template's parameter 
I would like to keep the function **_multiply_** to reduce the code because 5 places need to do the CAST.

The new function will look like below. If you have any idea please kindly let me know. Thank you.
```
        template <typename A, typename B, typename Accumulator>
        inline Accumulator multiply(A a, B b)
        {
            constexpr bool needAccumCast
                = !(std::is_same<A, Accumulator>() && std::is_same<B, Accumulator>())
                  && !std::is_same<A, Int8>() //case I8/I32/I32
                  && !std::is_same<A, Int8x4>(); //case I8x4/I32/I32

            using AMultT = std::conditional_t<needAccumCast, Accumulator, A>;
            using BMultT = std::conditional_t<needAccumCast, Accumulator, B>;
            return static_cast<Accumulator>(static_cast<AMultT>(a) * static_cast<BMultT>(b));
         }
```

Reply:
Ok, if we are to keep this function, then last thing I will request is from what @rosenrodt suggested above:

_Bonus: rearranging typename Accumulator to the leftmost side, the rest of them can even be deduced from the function arguments._

``` 
value += multiply<Accumulator>(a, b);    // A bit cleaner than value += multiply<TypeA, TypeB, Accumulator>(a, b); 
```

Reply:
Thanks for your review.
Updated the patch.